server/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -448,7 +448,9 @@ public NamedXContentRegistry getXContentRegistry() {
      */
     public final XContentParser contentParser() throws IOException {
         BytesReference content = requiredContent(); // will throw exception if body or content type missing
-        return xContentType.get().xContent().createParser(xContentRegistry, LoggingDeprecationHandler.INSTANCE, content.streamInput());
+        return xContentType.get().xContent().createParserForCompatibility(xContentRegistry,
+            LoggingDeprecationHandler.INSTANCE, content.streamInput(),
+            compatibleVersion == Version.CURRENT.minimumRestCompatibilityVersion());
```

Comment:
>  compatibleVersion == Version.CURRENT.minimumRestCompatibilityVersion());

I think it is a bit clean to:
```java
if(compatibleVersion == Version.CURRENT.minimumRestCompatibilityVersion())){
      return xContentType.get().xContent().createParserForCompatibility(xContentRegistry,
            LoggingDeprecationHandler.INSTANCE, content.streamInput());
else {
     return xContentType.get().xContent().createParser(xContentRegistry,
            LoggingDeprecationHandler.INSTANCE, content.streamInput());
}
```

It allows us to remove the isCompatible parameter on the "forComptabile" method, and just hard code to `true` when creating the new Parsers in the implementations, e.g. 
```java
return new YamlXContentParser(xContentRegistry, deprecationHandler, yamlFactory.createParser(is), true);
```